Convert existing modules to RST format and add labs where needed (#2)
AdaCore/training_material: courses/additional_modules/multiple_inheritance.rst

**********************
Multiple Inheritance
**********************

==============
Introduction
==============

------------------------------------------
Multiple Inheritance Is Forbidden In Ada
------------------------------------------

* There are potential conflicts with multiple inheritance

> * Some languages allow it:ambiguities have to be resolved when entities are referenced

Reviewer: ```suggestion
* Some languages allow it: ambiguities have to be resolved when entities are referenced
```

* Ada forbids it to improve integration
    
.. code:: Ada
    
    type Graphic is tagged record
       X, Y : Float;
    end record;
    function Get_X (V : Graphic) return Float;
       
    type Shape is tagged record
       X, Y : Float;
    end record;
    function Get_X (V : Shape) return Float;
       
    type Displayable_Shape is new Shape and Graphic with ...

----------------------------------

> Multiple Inheritance - Safe Case

Reviewer: ```suggestion
Multiple Inheritance |mdash| Safe Case
```
**Note**: `|mdash|` is only available when you include `isopub.txt` (using `.. include:: <isopub.txt>`).

Author: Pandoc does not give us "mdash" for free, so I'm going to leave this as-is for now

      type Graphic is interface;
      function Get_X (V : Graphic) return Float is abstract;
   
      type Shape is tagged record
         X, Y : Float;
      end record;
      function Get_X (V : Shape) return Float;
   
      type Displayable_Shape is new Shape and Graphic with ...

============
Interfaces
============

--------------------
Interfaces - Rules
--------------------

* An interface is a tagged type marked interface, containing

   - Abstract primitives
   - Null primitives
   - No fields

* Null subprograms provide default empty bodies to primitives that can be overridden

   .. code:: Ada

      type I is interface;
      procedure P1 (V : I) is abstract;
      procedure P2 (V : access I) is abstract
      function F return I is abstract;
      procedure P3 (V : I) is null;
 
* Note: null can be applied to any procedure (not only used for interfaces)

----------------------
Interface Derivation
----------------------

* An interface can be derived from another interface, adding primitives

   .. code:: Ada

      type I1 is interface;
      procedure P1 (V : I) is abstract;
      type I2 is interface and I1;
      Procedure P2 (V : I) is abstract;
 

> * A tagged type can derive from several interfacesand can derive from one interface several times

Reviewer: ```suggestion
* A tagged type can derive from several interfaces and can derive from one interface several times
```
